shared/actions/teams.js

```diff
@@ -648,40 +658,37 @@ function _checkRequestedAccessSuccess(result) {
 }
 
 const _saveChannelMembership = function(action: TeamsGen.SaveChannelMembershipPayload, state: TypedState) {
-  const {teamname, channelState} = action.payload
-  const convIDs: I.Set<ChatTypes.ConversationIDKey> = Constants.getTeamConvIDs(state, teamname)
-  const channelnameToConvID = keyBy(convIDs.toArray(), c => Constants.getChannelNameFromConvID(state, c))
+  const {teamname, oldChannelState, newChannelState} = action.payload
   const waitingKey = {key: `saveChannel:${teamname}`}
 
-  const calls = map(channelState, (wantsToBeInChannel: boolean, channelname: string) => {
-    if (wantsToBeInChannel) {
-      // $FlowIssue doesn't like callAndWrap
-      return Saga.callAndWrap(RPCChatTypes.localJoinConversationLocalRpcPromise, {
-        tlfName: teamname,
-        topicName: channelname,
-        topicType: RPCChatTypes.commonTopicType.chat,
-        visibility: RPCTypes.commonTLFVisibility.private,
-      })
+  const calls = []
+  for (const convIDStr in newChannelState) {
+    const convID = ChatTypes.stringToConversationIDKey(convIDStr)
+    if (oldChannelState[convID] === newChannelState[convID]) {
+      continue
     }
-    const convID =
-      channelnameToConvID[channelname] && ChatTypes.keyToConversationID(channelnameToConvID[channelname])
-    if (convID) {
-      // $FlowIssue doesn't like callAndWrap
-      return Saga.callAndWrap(RPCChatTypes.localLeaveConversationLocalRpcPromise, {
-        convID,
-      })
+
+    if (newChannelState[convID]) {
+      calls.push(
+        // $FlowIssue doesn't like callAndWrap
+        Saga.callAndWrap(RPCChatTypes.localJoinConversationByIDLocalRpcPromise, {
+          convID,
+        })
+      )
+    } else {
+      calls.push(
+        // $FlowIssue doesn't like callAndWrap
+        Saga.callAndWrap(RPCChatTypes.localLeaveConversationLocalRpcPromise, {
+          convID,
+        })
+      )
     }
-  }).filter(Boolean)
+  }
 
-  return Saga.all([
-    Saga.all(calls),
+  return Saga.sequentially([
```

Comment:
technically we can do the increment and the calls at the same time, but it's a bit clunky